plugins/woocommerce/src/Admin/SchemaBuilder/Types/AbstractSchemaType.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+namespace Automattic\WooCommerce\Admin\SchemaBuilder\Types;
+
+use Automattic\WooCommerce\Admin\SchemaBuilder\Types\SchemaReference;
+
+abstract class AbstractSchemaType {
+
+    /**
+     * Format.
+     *
+     * @var string|null
+     */
+    protected $format = null;
+
+    /**
+     * Required.
+     *
+     * @var bool
+     */
+    protected $required = false;
+
+    /**
+     * Description.
+     *
+     * @var string|null
+     */
+    protected $description = null;
+
+    /**
+     * Read only.
+     *
+     * @var bool
+     */
+    protected $readonly = false;
+
+    /**
+     * Validation callback.
+     *
+     * @var string
+     */
+    protected $callback = null;
+
+    /**
+     * Validation callback args.
+     *
+     * @var array
+     */
+    protected $callback_args = array();
+
+    /**
+     * Context.
+     *
+     * @var array
+     */
+    protected $context = array( 'view', 'edit' );
+
+    /**
+     * Schema type.
+     *
+     * @return string
+     */
+    abstract function get_type();
+
+    /**
+     * Set a description.
+     *
+     * @param string $description Description.
+     * @return SchemaTypeInterface
+     */
+    public function description( $description ) {
+        $this->description = $description;
+        return $this;
+    }
+
+    /**
+     * Set the property as required.
+     *
+     * @return SchemaTypeInterface
+     */
+    public function required() {
+        $this->required = true;
+        return $this;
+    }
+
+    /**
+     * Set the property as required.
+     *
+     * @return SchemaTypeInterface
+     */
+    public function readonly() {
+        $this->readonly = true;
+        return $this;
+    }
+
+    /**
+     * Context
+     */
+    public function context( $context ) {
+        $this->context = $context;
+        return $this;
+    }
+
+    /**
+     * Get the JSON.
+     */
+    public function get_json() {
+        $json = array(
+            'type'        => $this->get_type(),
+        );
+
+        if ( $this->description ) {
+            $json['description'] = $this->description;
+        }
+
+        if ( $this->format ) {
+            $json['format'] = $this->format;
+        }
+
+        if ( $this->context ) {
+            $json['context'] = $this->context;
+        }
+
+        if ( $this->readonly ) {
+            $json['readonly'] = true;
+        }
+
+        if ( $this->callback ) {
+            $json['$filters'] = array(
```

Comment:
`$filters` why did you go with this filters naming?
Is it necessary to surface these in the response? What is the use case you are thinking about?

Reply:
Opis is likely the JSON Schema validation tool we would continue to use since it's already a dependency of the remote specs validation.

`$filter` is [Opis's keyword for a callback function](https://opis.io/json-schema/2.x/php-filter.html), but this of course will change if we opted to use another validation library.